File: recipes/subms-spsc-ring-buffer/java/src/main/java/com/submillisecond/recipes/spsc/features/MpscFanIn.java
package com.submillisecond.recipes.spsc.features;

import com.submillisecond.recipes.spsc.SpscRingBuffer;

import java.util.ArrayList;
import java.util.List;

/**
 * N-producer single-consumer fan-in over N independent SPSC rings.
 *
 * <p>Each producer pushes into its own SPSC ring (so it stays wait-free
 * against its own counter), and the consumer round-robins across all rings.
 * The consumer never blocks any producer; producers never contend with each
 * other.
 *
 * <p>Memory cost: {@code N * capacity} slots. Round-robin is fair under
 * steady-state load; under skewed load the consumer spends extra
 * {@code tryPop} calls on quiet producers but never starves a busy one.
 */
public final class MpscFanIn<T> {

    private final List<SpscRingBuffer<T>> rings;
    private final List<SpscRingBuffer<T>.Producer> producers;
    private final List<SpscRingBuffer<T>.Consumer> consumers;
    private final Consumer consumer;

    /** Build with {@code producerCount} rings of capacity {@code perRingCapacity}. */
    public MpscFanIn(int producerCount, int perRingCapacity) {
        if (producerCount < 1) throw new IllegalArgumentException("need at least one producer");
        this.rings = new ArrayList<>(producerCount);
        this.producers = new ArrayList<>(producerCount);
        this.consumers = new ArrayList<>(producerCount);
        for (int i = 0; i < producerCount; i++) {
            SpscRingBuffer<T> r = new SpscRingBuffer<>(perRingCapacity);
            rings.add(r);
            producers.add(r.producer());
            consumers.add(r.consumer());
        }
        this.consumer = new Consumer();
    }

    public int producerCount() { return rings.size(); }

    /** Producer handle at index {@code i}. Move it to its owning thread. */
    public Producer producer(int i) { return new Producer(producers.get(i)); }

    /** The single consumer that round-robins across all producer rings. */
    public Consumer consumer() { return consumer; }

    /** One producer side. Wait-free against its own ring. */
    public final class Producer {
        private final SpscRingBuffer<T>.Producer inner;
        private Producer(SpscRingBuffer<T>.Producer inner) { this.inner = inner; }

        public boolean tryPush(T value) { return inner.tryPush(value); }
    }

    /** The consumer side. Round-robins across producer rings. */
    public final class Consumer {
        private int cursor;

        /**
         * Probe each producer ring once starting at the current cursor. Returns
         * the first value found, advancing the cursor past the producing ring
         * so the next call starts at a fresh point.
         */
        public T tryPop() {
            int n = consumers.size();
            for (int offset = 0; offset < n; offset++) {
                int idx = (cursor + offset) % n;
                T v = consumers.get(idx).tryPop();
                if (v != null) {
                    cursor = (idx + 1) % n;
                    return v;
                }
            }
            return null;
        }

        public int producerCount() { return consumers.size(); }
    }
}

File: recipes/subms-spsc-ring-buffer/rust/src/features/mpsc_fan_in.rs
//! N-producer single-consumer fan-in over N independent SPSC rings.
//!
//! Each producer pushes into its own SPSC ring (so it stays wait-free against
//! its own counter), and the consumer round-robins across all rings. The
//! consumer never blocks any producer; producers never contend with each
//! other.
//!
//! Memory cost: `N * capacity` slots. Round-robin is fair under steady-state
//! load; under skewed load the consumer spends extra `try_pop` calls on
//! quiet producers but never starves a busy one.

use crate::{Consumer, Producer, SpscRingBuffer};

/// Builder + handle factory for an N-producer fan-in. After construction,
/// move each `MpscFanInProducer` to its producing thread and the single
/// `MpscFanInConsumer` to its consuming thread.
pub struct MpscFanIn;

impl MpscFanIn {
    /// Build `producer_count` SPSC rings of capacity `per_ring_capacity` and
    /// return matched producer / consumer handles.
    pub fn with_capacity<T: Send + 'static>(
        producer_count: usize,
        per_ring_capacity: usize,
    ) -> (Vec<MpscFanInProducer<T>>, MpscFanInConsumer<T>) {
        assert!(producer_count >= 1, "need at least one producer");
        let mut producers = Vec::with_capacity(producer_count);
        let mut consumers = Vec::with_capacity(producer_count);
        for _ in 0..producer_count {
            let (p, c) = SpscRingBuffer::with_capacity::<T>(per_ring_capacity);
            producers.push(MpscFanInProducer { inner: p });
            consumers.push(c);
        }
        (
            producers,
            MpscFanInConsumer {
                rings: consumers,
                cursor: 0,
            },
        )
    }
}

/// One producer side of an `MpscFanIn`. Wait-free against its own ring;
/// independent of any other producer.
pub struct MpscFanInProducer<T> {
    inner: Producer<T>,
}

impl<T> MpscFanInProducer<T> {
    /// Push a value into this producer's ring. Returns `Err(value)` if full.
    pub fn try_push(&mut self, value: T) -> Result<(), T> {
        self.inner.try_push(value)
    }

    pub fn capacity(&self) -> usize {
        self.inner.capacity()
    }
}

/// The single consumer side. Round-robins across the producer rings.
pub struct MpscFanInConsumer<T> {
    rings: Vec<Consumer<T>>,
    /// Next ring to probe; advances every successful pop so producers don't
    /// starve. Wraps modulo `rings.len()`.
    cursor: usize,
}

impl<T> MpscFanInConsumer<T> {
    /// Try one full round of probes across all producer rings. Returns the
    /// first value found, advancing the cursor past the producing ring so
    /// the next call starts at a fresh point.
    pub fn try_pop(&mut self) -> Option<T> {
        let n = self.rings.len();
        for offset in 0..n {
            let idx = (self.cursor + offset) % n;
            if let Some(v) = self.rings[idx].try_pop() {
                self.cursor = (idx + 1) % n;
                return Some(v);
            }
        }
        None
    }

    /// Producer count.
    pub fn producer_count(&self) -> usize {
        self.rings.len()
    }
}

#[cfg(test)]
#[path = "mpsc_fan_in_tests.rs"]
mod tests;
